include/lsst/afw/image/TransmissionCurve.h

```diff
@@ -0,0 +1,300 @@
+
+    //@{
+    /**
+     *  Evaluate the throughput at a position into a provided output array.
+     *
+     *  @param[in]  position     Spatial position at which to evaluate.
+     *  @param[in]  wavelengths  Wavelengths at which to evaluate.
+     *
+     *  @param[in,out]  out      Computed throughput values.  Must be pre-
+     *                           allocated to the same size as the wavelengths
+     *                           array.  May include NaN values if the
+     *                           requested wavelengths are beyond bounds
+     *                           reported by getWavelengthBounds(). Out-of-
+     *                           bounds wavelengths may also yield zero
+     *                           throughput values if throughput is known to
+     *                           be limited to those bounds.
+     *
+     *  @throws pex::exceptions::DomainError if the given position is outside
+     *          the region for which the TransmissionCurve is defined.
+     */
+    virtual void sampleAt(
+        geom::Point2D const & position,
+        ndarray::Array<double const,1,1> const & wavelengths,
+        ndarray::Array<double,1,1> const & out
+    ) const = 0;
+
+    void sampleAt(
+        geom::Point2D const & position,
+        SampleDef const & wavelengths,
+        ndarray::Array<double,1,1> const & out
+    ) const;
+    //@}
+
+
+    //@{
+    /**
+     *  Evaluate the throughput at a position into a new array.
+     *
+     *  @param[in]  position     Spatial position at which to evaluate.
+     *  @param[in]  wavelengths  Wavelengths at which to evaluate.
+     *
+     *  @return  Computed throughput values, in an array with the same size as
+     *           wavelengths.  May include NaN values if the requested
+     *           wavelengths are beyond bounds reported by
+     *           getWavelengthBounds().  Out-of-bounds wavelengths may also
+     *           yield zero throughput values if throughput is known to be
+     *           limited to those bounds.
+     *
+     *  @throws pex::exceptions::DomainError if the given position is outside
+     *          the region for which the TransmissionCurve is defined.
+     */
+    ndarray::Array<double,1,1> sampleAt(
+        geom::Point2D const & position,
+        ndarray::Array<double const,1,1> const & wavelengths
+    ) const;
+
+    ndarray::Array<double,1,1> sampleAt(
+        geom::Point2D const & position,
+        SampleDef const & wavelengths
+    ) const;
+    //@}
+
+protected:
+
+    /**
+     *  Polymorphic implementation for transform().
+     *
+     *  The default implementation of this method creates a new
+     *  TransmissionCurve that lazily applies the given transform to points
+     *  before evaluating the original TransmissionCurve, which should be
+     *  appropriate for nearly all concrete TransmissionCurve subclases.
+     *
+     *  @param[in] self       A shared_ptr to this.
```

Comment:
So the implementation can return a `shared_ptr` to this without the overhead that comes with (mutiple) inheritance from `std::enable_shared_ptr_from_this`.

Reply:
(but yes, I should document that, since it's a very reasonable question that anyone seeing this code will ask)

Reply:
Can you elaborate? This seems like exactly the situation `enabled_shared_from_this` was designed for (Effective Modern C++ even recommends using factory methods to ensure that all objects of the class have an associated `shared_ptr`).

Reply:
I believe `enable_shared_from_this` adds a `weak_ptr` data member.  It's certainly also an option, but it's an unnecessary step since it can be avoided by adding the extra argument to a completely hidden member function.

Reply:
Except that in C++, it's extremely unidiomatic, and therefore confusing to anybody working on the code in the future.

I'll need to check tomorrow on exactly what `enable_shared_from_this` does, but I was under the impression that it's only a class because it's handy for it to act like a functor.

Reply:
Actually, `enable_shared_from_this` has the added bonus that it lets me make the public `transform` method a regular bound method, and that definitely tips me in its favor.  I'll go with that.

Reply:
Naive question, but enable_shared_from_this is only safe if there is an existing shared_ptr to this. Will users safely be able to assume that here?

Reply:
@r-owen it's safe unless somebody makes a subclass of `TransmissionCurve` with a public constructor. Unfortunately in C++ there's no way to guard against that.